### PR TITLE
Purge non-empty nulls in `cudf::make_lists_column`

### PR DESCRIPTION
**Pull request by ttnghia**
During construction, the new lists column is also applied a null mask. Such null superposition allows the existence of non-empty nulls, which may undefined behavior later on for various cudf APIs.

This PR fixes it by removing non-empty nulls after lists column constructions. It also fixes unused variables `stream` and `mr`.

After lists creation is corrected in this PR, a lot of tests failed because either these tests were incorrect, or the corresponding APIs were wrongly handling nulls (i.e., they generate outputs with non-empty null lists). Below is the list of the tests that are also fixed in this PR:
 - [X] COPYING_TEST
 - [X] LISTS_TEST
 - [X] PARQUET_TEST
 - [X] STRUCTS_TEST
 - [X] UTILITIES_TEST

In addition, `JSON_TEST` failed due to an unrelated bug (https://github.com/rapidsai/cudf/issues/12418). The bug already existed there but is discovered only now due to changes in this PR. The corresponding failed test is just temporarily disabled and will be fixed separately.

Also closes:
 * https://github.com/rapidsai/cudf/issues/12405

### CODE REVIEW

**Comment by ttnghia** [comment]
After lists creation is corrected in this PR, a lot of tests failed because either these tests were incorect, or the corresponding APIs were wrongly handling nulls (i.e., they generate outputs with non-empty null lists):
 * COPYING_TEST
 * JSON_TEST
 * LISTS_TEST
 * PARQUET_TEST
 * STRUCTS_TEST
 * TRANSFORM_TEST
 * UTILITIES_TEST

Taking this into draft, waiting to fix these APIs.

File: cpp/src/lists/lists_column_factories.cu [hunk 1]
```diff
@@ -107,12 +108,15 @@ std::unique_ptr<column> make_lists_column(size_type num_rows,
...
-                                  std::move(children));
+
+  auto output = std::make_unique<column>(cudf::data_type{type_id::LIST},
+                                         num_rows,
+                                         rmm::device_buffer{},
+                                         std::move(null_mask),
+                                         null_count,
+                                         std::move(children));
+  return null_count == 0 ? std::move(output)
+                         : detail::purge_nonempty_nulls(output->view(), stream, mr);
```
Comment (ttnghia) [inline cpp/src/lists/lists_column_factories.cu:119]: This is the main change of this entire PR.

File: cpp/tests/column/factories_test.cpp [hunk 2]
```diff
@@ -761,6 +762,41 @@ TEST_F(ListsZeroLengthColumnTest, MixedTypes)
   }
 }
 
+TEST_F(ListsZeroLengthColumnTest, SuperimposeNulls)
```
Comment (ttnghia) [inline cpp/tests/column/factories_test.cpp:765]: This is a new test for the change in this PR.

File: cpp/tests/io/json_test.cpp [hunk 3]
```diff
@@ -1463,7 +1463,7 @@ TEST_P(JsonReaderParamTest, JsonDtypeSchema)
                                  cudf::test::strings_column_wrapper({"aa ", "  bbb"}));
 }
 
-TEST_F(JsonReaderTest, JsonNestedDtypeSchema)
+TEST_F(JsonReaderTest, DISABLED_JsonNestedDtypeSchema)
```
Comment (ttnghia) [inline cpp/tests/io/json_test.cpp:1466]: Disable due to https://github.com/rapidsai/cudf/issues/12418.

File: cpp/src/lists/lists_column_factories.cu [hunk 4]
```diff
@@ -104,15 +105,25 @@ std::unique_ptr<column> make_lists_column(size_type num_rows,
...
+                                         rmm::device_buffer{},
+                                         std::move(null_mask),
+                                         null_count,
+                                         std::move(children));
+
+  // We need to enforce all null lists to be empty.
+  // Checking `null_count==0` is not enough as it can be `UNKNOWN_NULL_COUNT` for nullable column.
+  return null_count == 0 || !output->nullable() || child_type_id == type_id::EMPTY
+           ? std::move(output)
+           : detail::purge_nonempty_nulls(output->view(), stream, mr);
```
Comment (hyperbolic2346) [inline cpp/src/lists/lists_column_factories.cu:126]: Are there concerns about the performance implications here where we have gone from a move to a copy in the case of calling `purge_nonempty_nulls`?

File: cpp/tests/lists/combine/concatenate_rows_tests.cpp [hunk 5]
```diff
@@ -814,17 +814,16 @@ TEST_F(ListConcatenateRowsNestedTypesTest, StructWithNulls)
 
     // expected
     cudf::test::fixed_width_column_wrapper<int> se_0{
-      {10, 11, 12, 13, 2, 3, 4, 14, 15, 16, 5, 17, 18, 19, 20, 21, 22, 23, 24, 6, 7},
-      nulls_at({18})};
+      {10, 11, 12, 13, 2, 3, 4, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 6, 7}, nulls_at({17})};
     cudf::test::strings_column_wrapper se_1{
-      {"arg",    "mno",     "ampere", "gpu",   "bananas", "",      "",    "",    "hhh", "warp", "g",
+      {"arg",    "mno",     "ampere", "gpu",   "bananas", "",      "",    "",    "hhh", "warp",
```
Comment (hyperbolic2346) [inline cpp/tests/lists/combine/concatenate_rows_tests.cpp:819]: Why was this change made?

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
A few questions for this one mostly.

File: cpp/src/lists/lists_column_factories.cu (hunk 4 above)
Comment (ttnghia) [reply cpp/src/lists/lists_column_factories.cu:126]: `purge_nonempty_nulls` is a very trivial gathering/memcpy thus it should have very small performance impact.

File: cpp/tests/lists/combine/concatenate_rows_tests.cpp (hunk 5 above)
Comment (ttnghia) [reply cpp/tests/lists/combine/concatenate_rows_tests.cpp:819]: At line 762 above, the `"g"` string corresponds to a null list (line 769, the last null). Before this PR, that null list still store `"g"` which will be concatenated later to the final output. With this PR, all null lists are empty thus `"g"` is removed completely.

**Review by hyperbolic2346 — APPROVED** [review]

**Pull request merged**